subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,264 @@
+	// while in the production targets the corresponding target is marked as `prod`.
+	// TODO: switch to usage of /api/v1/user_repo/targets.json API once it supports `production` param
+	// instead of doing the stuff below
+	custom, err := getTargetCustomInfo(targetName)
+	subcommands.DieNotNil(err)
+
+	if !prod && custom.Tags[0] != tag {
+		fmt.Fprintf(os.Stderr, "The specified Target hasn't been found; target: %s, tag: %s\n",
+			targetName, tag)
+		os.Exit(1)
+	}
+
+	if prod {
+		isProdTarget, err := checkIfProdTarget(targetName)
+		subcommands.DieNotNil(err)
+		if !isProdTarget {
+			fmt.Fprintf(os.Stderr, "The specified Target is not a production target; target %s, tag: %s\n",
+				targetName, tag)
+			os.Exit(1)
+		}
+	}
+
+	targetVer, err := strconv.Atoi(custom.Version)
+	subcommands.DieNotNil(err)
+
+	fmt.Println("Downloading TUF metadata...")
+	err = downloadTufRepo(factory, tag, prod, path.Join(dstDir, "tuf"))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to download TUF repo: %s", err.Error())
+		os.Exit(1)
+	}
+
+	hardwareID := custom.HardwareIds[0]
+	buildTag := custom.Tags[0]
+	targetOstreeVer := targetVer
+	if len(custom.OrigUri) > 0 {
+		indx := strings.LastIndexByte(custom.OrigUri, '/')
+		if indx == -1 {
+			fmt.Fprintf(os.Stderr, "Failed to determine Target's OE build version: %s", custom.OrigUri)
+			os.Exit(1)
+		}
+		targetOstreeVerStr := custom.OrigUri[indx+1:]
+		targetOstreeVer, err = strconv.Atoi(targetOstreeVerStr)
+		subcommands.DieNotNil(err)
+	}
+	fmt.Printf("Downloading an ostree repo from the Target's OE build %d...\n", targetOstreeVer)
+	err = downloadOstree(targetOstreeVer, hardwareID, dstDir)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to download Target ostree repo: %s", err.Error())
+		os.Exit(1)
+	}
+
+	fmt.Printf("Downloading Apps fetched by the `assemble-system-image` run of the build:  %d...\n", targetVer)
+	err = downloadApps(targetName, targetVer, buildTag, path.Join(dstDir, "apps"))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to download Target Apps: %s", err.Error())
+		os.Exit(1)
+	}
+	fmt.Println("Successfully downloaded offline update content")
+}
+
+func checkIfProdTarget(targetName string) (bool, error) {
+	meta, err := api.ProdTargetsGet(factory, tag, false)
+	if err != nil {
+		return false, err
+	}
+	if meta == nil {
+		return false, nil
+	}
+	prodTargets := meta.Signed.Targets
+	for tn := range prodTargets {
+		if tn == targetName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func getTargetCustomInfo(targetName string) (*client.TufCustom, error) {
+	targetFile, err := api.TargetGet(factory, targetName)
+	if err != nil {
+		return nil, err
+	}
+	custom, err := api.TargetCustom(*targetFile)
+	return custom, err
+}
+
+func downloadTufRepo(factory string, tag string, prod bool, dstDir string) error {
+	ver := 1 // TODO: start from the second version once the first one is provisioned on a device during manufacturing process.
```

Comment:
> and let the device decide which ones it wants to take from the offline source.

Yep, although downloading and carrying 1.root.json on a flash disk looks like a security flaw to me.

Also, during manufacturing process, the initial root version must be provisioned starting from the first version, otherwise aklite will fail.

Reply:
IMHO if we're talking about security here - then both 1.root.json and 2.root.json (I'd say 3.root.json too) should be pre-provisioned, as version 2 is the first one we trust, and version 3 is the first one customer can trust:
- 1.root.json - auto-generated by tuf-reposerver;
- 2.root.json - we took root and targets role keys away from tuf-reposerver i.e. generated new online keys;
- 3.root.json - the customer took the root keys offline (the first secure version from the customer perspective).